### PR TITLE
More generic contract generation

### PR DESCRIPTION
**Pull request by vkleen**
This PR introduces a type variable environment into labels and makes the tracking type variable polarities for contract generation dynamic. Additionally, it introduces two new primops to manipulate this environment.

This is preparatory work for #1177.

### CODE REVIEW

File: src/eval/operation.rs [hunk 1]
```diff
@@ -3106,6 +3139,92 @@ impl<R: ImportResolver, C: Cache> VirtualMachine<R, C> {
...
+                        RichTerm {
+                            term: polarity,
+                            pos: pos2,
+                        },
+                    ))
+                };
+
+                let Term::Lbl(label) = &*label else {
+                    return Err(EvalError::TypeError(
+                        String::from("Lbl"),
```
Comment (yannham) [inline src/eval/operation.rs:3208]:
```suggestion
                        String::from("Label"),
```

File: src/label.rs [hunk 2]
```diff
@@ -366,8 +371,42 @@ pub struct Label {
...
     pub path: ty_path::Path,
+    /// An environment mapping type variables to [`TypeVarData`]. Used by
+    /// polymorphic contracts to decide which actions to take when encountering a `forall`.
+    pub type_environment: HashMap<SealingKey, TypeVarData>,
 }
 
+/// Data about type variables that is needed for polymorphic contracts to decide which actions to take.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct TypeVarData {
+    pub polarity: Polarity,
```
Comment (yannham) [inline src/label.rs:382]: I guess this field will take on more values once you do the "one-pass" version? Morally, `positive`, `negative`, and the composite partial identities `(positive, negative)` and `(negative, positive)` ?

File: src/label.rs [hunk 3]
```diff
@@ -366,8 +371,42 @@ pub struct Label {
...
+                Field::from(RichTerm::from(Term::from(value.polarity))),
+            )]
+            .into(),
+            attrs: Default::default(),
+            sealed_tail: None,
+        })
+    }
+}
+
+impl HasUnifType for TypeVarData {
```
Comment (yannham) [inline src/label.rs:399]: I find the name a tad misleading, I would expect a trait named `HasUnifType` to define a method `has_unif_type(&self) -> bool`. I don't know of a good name scheme for those kind of `* -> *` functions from a type to a type. `AssociatedUnifType`? Looks a bit long. `ReifyAsUnifType`? `ReifyAsNickelType`? I like reify, it sounds professional :sweat_smile:

File: stdlib/internals.ncl [hunk 4]
```diff
@@ -25,9 +25,10 @@
       else
           %blame% label,
 
-  "$forall_var" = fun sealing_key polarity label value =>
-      let current-polarity = %polarity% label in
-      if polarity == current-polarity then
+  "$forall_var" = fun sealing_key label value =>
+      let current_polarity = %polarity% label in
+      let {polarity, ..} = %lookup_type_variable% sealing_key label in
```
Comment (yannham) [inline stdlib/internals.ncl:30]: I don't know if it makes a big difference, but destructuring involves a contract application and thus merging. It's a bit less visible, but I would use a more down-to-earth `let var_data = ...` and `var_data.polarity` in such low-level code, which is not very pretty anyway, to be avoid the codepath of merging to be taken.

But honestly it couldn't also not make any noticeable difference at all, I'm not sure.

```suggestion
      let {polarity, ..} = %lookup_type_variable% sealing_key label in
```

**Review by yannham — APPROVED** [review]
It's not urgent, but looking at code each time we add a new primop, it really feels cumbersome to do (add stuff to the lexer, to the parser, implement boiler plate code for dynamically typing the arguments, extract them, oh should I take ownership with `match_shared!` or not :man_shrugging: ). It's not urgent but there must be a better way of organizing (and splitting, as well) `eval::operation`.

File: src/label.rs (hunk 2 above)
Comment (vkleen) [reply src/label.rs:382]: I was expecting to need more data for type variables, indeed. Originally I was also tracking the variable kind, but that ended up being unnecessary for now.

File: src/label.rs (hunk 3 above)
Comment (vkleen) [reply src/label.rs:399]: +1 for reify 👑 Another possibility would be `AsUnifType`. Really, the problem seems to be that `TypeVarData::unif_type()` should actually be an associated constant; and that the trait can't be named `UnifType` to conform to the usual naming format.

Comment (yannham) [reply src/label.rs:399]: Right, I would also have called it `UnifType` naturally.

Comment (vkleen) [reply src/label.rs]: I've renamed it to `ReifyAsUnifType`. Yell if that's a bad idea 😅

**Pull request merged**